samples/template/src/main.c

```diff
@@ -0,0 +1,517 @@
+		(SID_STATUS_TIME_SYNCED == status->detail.time_sync_status) ? "Success" : "Fail",
+		(status->detail.link_status_mask & SID_LINK_TYPE_1) ? "Up" : "Down",
+		(status->detail.link_status_mask & SID_LINK_TYPE_2) ? "Up" : "Down",
+		(status->detail.link_status_mask & SID_LINK_TYPE_3) ? "Up" : "Down");
+
+	for (int i = 0; i < SID_LINK_TYPE_MAX_IDX; i++) {
+		enum sid_link_mode mode =
+			(enum sid_link_mode)status->detail.supported_link_modes[i];
+
+		if (mode) {
+			LOG_INF("Link mode on %s = {Cloud: %s, Mobile: %s}",
+				(SID_LINK_TYPE_1_IDX == i) ? "BLE" :
+				(SID_LINK_TYPE_2_IDX == i) ? "FSK" :
+				(SID_LINK_TYPE_3_IDX == i) ? "LoRa" :
+							     "unknow",
+				(mode & SID_LINK_MODE_CLOUD) ? "True" : "False",
+				(mode & SID_LINK_MODE_MOBILE) ? "True" : "False");
+		}
+	}
+}
+
+static void state_running_run(void *o)
+{
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+	sid_error_t e = SID_ERROR_NONE;
+
+	switch (ctx->sm.event) {
+	case APP_EVENT_SIDEWALK:
+		e = sid_process(ctx->handle);
+		if (e) {
+			LOG_ERR("sid process err %d", (int)e);
+		}
+		break;
+	case APP_EVENT_FACTORY_RESET:
+		e = sid_set_factory_reset(ctx->handle);
+		if (e) {
+			LOG_ERR("sid process err %d", (int)e);
+		}
+		break;
+	case APP_EVENT_ERROR:
+		LOG_ERR("sid state err %d", (int)sid_get_error(ctx->handle));
+		break;
+	case APP_EVENT_LINK_SWITCH:
+		sid_error_t e = SID_ERROR_NONE;
+
+		static uint32_t link_mask = SID_LINK_TYPE_1;
+
+		switch (link_mask) {
+		case SID_LINK_TYPE_1:
+			link_mask = SID_LINK_TYPE_2;
+			break;
+		case SID_LINK_TYPE_2:
+			link_mask = SID_LINK_TYPE_3;
+			break;
+		default:
+			link_mask = SID_LINK_TYPE_1;
+			break;
+		}
+
+		LOG_INF("Sidewalk link switch to 0x%x", link_mask);
+		ctx->config.link_mask = link_mask;
+
+		if (ctx->handle != NULL) {
+			e = sid_deinit(ctx->handle);
+			if (e) {
+				LOG_ERR("sid deinit err %d", (int)e);
+			}
+		}
+
+		e = sid_init(&ctx->config, &ctx->handle);
+		if (e) {
+			LOG_ERR("sid init err %d", (int)e);
+		}
+
+		e = sid_start(ctx->handle, link_mask);
+		if (e) {
+			LOG_ERR("sid start err %d", (int)e);
+		}
+		break;
+	case APP_EVENT_READY:
+	case APP_EVENT_NOT_READY:
+		break;
+	default:
+		LOG_WRN("event: unknow %d", ctx->sm.event);
+		break;
+	}
+}
+
+static void state_init_entry(void *o)
+{
+	LOG_INF("state: init (entry)");
+
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+	sid_error_t e = SID_ERROR_NONE;
+
+	e = application_pal_init();
+	if (e) {
+		LOG_ERR("pal init err %d", (int)e);
+	}
+	e = sid_init(&ctx->config, &ctx->handle);
+	if (e) {
+		LOG_ERR("sid init err %d", (int)e);
+	}
+	e = sid_start(ctx->handle, SID_LINK_TYPE_1);
+	if (e) {
+		LOG_ERR("sid start err %d", (int)e);
+	}
+}
+
+static void state_init_run(void *o)
+{
+	LOG_INF("state: init");
+
+	struct sid_ctx_s *ctx = CONTAINER_OF(o, struct sid_ctx_s, sm);
+
+	switch (ctx->sm.event) {
+	case APP_EVENT_SIDEWALK:
```

Comment:
AS this is the application state, it should not care about the events from the outer SMF. Using separate enum for application events will give us a neat way of ignoring values of events that are related to outer SMF

Reply:
As before: I'm not sure if zephyr state machine allows to have a parent form different state machine instance. 
For now there is no inner and outer state machine, but machine with a parent state. And the common events are handled in parent state to not repeat the code. But we consider implementing this like you said, worth to check.